AzureMonitorAgent/agent.py

```diff
@@ -1087,6 +1084,79 @@ def exit_if_vm_not_supported(operation):
     return 0
 
 
+def is_arc_installed():
+    """
+    Check if this is an Arc machine
+    """
+    # Using systemctl to check this since Arc only supports VMs that have systemd
+    check_arc = os.system('systemctl status himdsd 1>/dev/null 2>&1')
+    return check_arc == 0
+
+
+def get_arc_endpoint():
+    """
+    Find the endpoint for Arc IMDS
+    """
+    endpoint_filepath = '/lib/systemd/system.conf.d/azcmagent.conf'
+    endpoint = ''
+    try:
+        with open(endpoint_filepath, 'r') as f:
+            data = f.read()
+        endpoint = data.split("\"IMDS_ENDPOINT=")[1].split("\"\n")[0]
+    except:
+        hutil_log_error('Unable to load Arc IMDS endpoint from {0}'.format(endpoint_filepath))
+    return endpoint
+
+
+def get_imds_endpoint():
+    """
+    Find the appropriate endpoint (Azure or Arc) for IMDS
+    """
+    azure_imds_endpoint = 'http://169.254.169.254/metadata/instance?api-version=2018-10-01'
```

Comment:
If HTTP request to this URL is to be executed outside of Azure Networking, it might be easy enough to spoof it and feed something into the extension that's not expected, we may need to add some input validation next time around for basics like string length limits and so on for extra protection.

Reply:
IMDS actually comes from the host node, so HTTP isn't an issue (in fact, IMDS won't even respond over HTTPS). [Docs](https://docs.microsoft.com/en-us/azure/virtual-machines/windows/instance-metadata-service?tabs=linux):
```
IMDS is a REST API that's available at a well-known, non-routable IP address (169.254.169.254). You can only access it from within the VM. Communication between the VM and IMDS never leaves the host. Have your HTTP clients bypass web proxies within the VM when querying IMDS, and treat 169.254.169.254 the same as 168.63.129.16.
```

Reply:
My worry is that holds true for Azure VMs only where MS controls networking (but I'm not certain). If AMA is on non-azure VM (via ARC, or later without ARC), can it be routable/spoofable IP? Last time I was working on similar subject with SDK team, it turned out that SDK may try to reach it from inside the app on-prem or on laptop or device, so IP is not controlled by MS at this situation and we needed to do some protection. I'm trying to ensure we do not miss similar case for agent that might be running off-Azure controlled network.

Reply:
On Arc, Arc IMDS comes from localhost so I think it should be ok (in the sense that if someone is spoofing IMDS on the box, there are bigger problems at hand). I wasn't aware of plans to make this work in non-Azure/non-Arc scenario, but seems like the kind of thing that would require significant re-working of IMDS-reliant logic regardless. 

If we did want to add protection, what options do we have? We can't use HTTPS because IMDS will not respond to that.

Reply:
Makes sense for ARC, if ARC is not there, extension won't be setup in the first place. I'm not sure on our architecture to support no-ARC machine and how we are going (if at all) to communicate to IMDS there, this particular script might not be executed at all then.
Generally (when time comes to it), in such a case (when we cannot SSL to endpoint), best we can do is protecting against bad input (basic input validation), ensuring that we won't pass too big of a string into unamanaged code and so on.